torchvision/csrc/io/image/cpu/decode_png.cpp

```diff
@@ -25,27 +25,41 @@ torch::Tensor decode_png(const torch::Tensor& data, ImageReadMode mode) {
   auto info_ptr = png_create_info_struct(png_ptr);
   if (!info_ptr) {
     png_destroy_read_struct(&png_ptr, nullptr, nullptr);
-    // Seems redundant with the if statement. done here to avoid leaking memory.
+    // Seems redundant with the if statement. Done here to avoid leaking memory.
     TORCH_CHECK(info_ptr, "libpng info structure allocation failed!")
   }
 
-  auto datap = data.accessor<unsigned char, 1>().data();
+  auto deleter = [&png_ptr, &info_ptr](png_structp __unused_ptr) {
+    png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
+  };
+  std::unique_ptr<png_struct, decltype(deleter)> png_ptr_RAII(png_ptr, deleter);
+
+  auto accessor = data.accessor<unsigned char, 1>();
+  auto datap = accessor.data();
+  auto data_size = accessor.size(0);
 
   if (setjmp(png_jmpbuf(png_ptr)) != 0) {
-    png_destroy_read_struct(&png_ptr, &info_ptr, nullptr);
     TORCH_CHECK(false, "Internal error.");
   }
-  auto is_png = !png_sig_cmp(datap, 0, 8);
+
+  auto is_png = (data_size >= 8 && !png_sig_cmp(datap, 0, 8));
```

Comment:
Nice!

We should probably check if we need to update https://github.com/pytorch/vision/blob/a83b9a17e441e6d77e9d59ca14d75fb5ba8c31f6/torchvision/csrc/io/image/cpu/decode_image.cpp#L20 to contain the full signature described [in the PNG specifications](https://www.w3.org/TR/2003/REC-PNG-20031110/#5PNG-file-signature)